### PR TITLE
Switch cudagraph backend to cudagraph trees

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121033
* __->__ #121019
* #121018
* #121017

Switch torch.compile(..., backend="cudagraphs") to use cudagraph trees. Enabled a few test in cudagraph_trees and note that there is another test suite existing for cudagraphs backend: https://github.com/pytorch/pytorch/blob/main/test/dynamo/test_cudagraphs.py.

This is basically the inductor cudagraphs without inductor.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang @ErnestChan 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/_inductor/utils.py [hunk 1]
```diff
@@ -680,6 +680,13 @@ def has_incompatible_cudagraph_ops(gm):
     return False
 
 
+def output_node(gm: torch.fx.GraphModule):
+    """Get the output node from an FX graph"""
+    last_node = next(iter(reversed(gm.graph.nodes)))
+    assert last_node.op == "output"
+    return last_node
```
Comment (ezyang) [inline torch/_inductor/utils.py:687]: @jansel weren't you telling us about how sometimes the output node isn't last lol

File: torch/__init__.py [hunk 2]
```diff
@@ -1767,6 +1767,10 @@ def __eq__(self, other):
     def __call__(self, model_, inputs_):
         return self.compiler_fn(model_, inputs_, **self.kwargs)
 
+    def reset(self):
+        if hasattr(self.compiler_fn, "reset"):
```
Comment (ezyang) [inline torch/__init__.py:1771]: Do you really want to swallow this call entirely if there is no reset on compiler_fn? Seems better to error no?

File: torch/_inductor/cudagraph_utils.py [hunk 3]
```diff
@@ -93,3 +94,12 @@ def check_lowering_disable_cudagraph(
     device_node_mapping: Dict[torch.device, torch.fx.Node]
 ):
     return check_multiple_devices_or_any_cpu_nodes(device_node_mapping)
+
+
+@dataclasses.dataclass
+class BoxedDeviceIndex:
+    value: Optional[int]
+
+    def set(self, device_idx):
```
Comment (ezyang) [inline torch/_inductor/cudagraph_utils.py:103]: nit: put a type annotation?

**Review by ezyang — APPROVED** [review]

File: torch/__init__.py (hunk 2 above)
Comment (eellison) [reply torch/__init__.py:1771]: Not all backend compilers define reset. I dont think its better to error.

Comment (ezyang) [reply torch/__init__.py:1771]: Are you unconditionally calling reset somewhere?

File: torch/_inductor/utils.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/utils.py:687]: I think others were saying these asserts fail.  I think we should enforce this constraint.

**Review by jansel — APPROVED** [review]

**Comment by xmfan** [comment]
I think this PR is causing segfaults on Dynamo exit, from bisecting https://github.com/pytorch/pytorch/issues/121862. Reverting it passes the repros for me locally. We're still investigating why this is green in CI

**Pull request closed without merging**